service/src/main/kotlin/fi/espoo/evaka/application/ApplicationControllerV2.kt

```diff
@@ -294,12 +294,11 @@ class ApplicationControllerV2(
                     personService.getGuardians(tx, user, application.childId)
                         .map { personDTO -> PersonJSON.from(personDTO) }
 
-                val attachments =
-                    if (accessControl.hasPermissionFor(user, Action.Application.READ_ATTACHMENTS, applicationId)) {
-                        tx.getApplicationAttachments(applicationId)
-                    } else {
-                        listOf()
+                val attachments = tx.getApplicationAttachments(applicationId).let { allAttachments ->
+                    allAttachments.filter { attachment ->
+                        accessControl.hasPermissionFor(user, Action.Attachment.READ_APPLICATION_ATTACHMENT, attachment.id)
```

Comment:
You should actually do this like @Gekkio suggested in the previous comment. Calling `accessControl.hasPermissionFor()` in a loop causes as many database queries as there are attachments. You can use `accessControl.checkPermissionFor` by giving it the ids of all attachments, and then act on the the `Map` it returns.